File: client/src/Services/AuthService.ts
import axios from "axios";
import { debug } from "console";
import { ApiUrl } from "./Data";
import Moment from 'moment';

const url = ApiUrl() + "auth/";

export const login = async (loginDto: { login: string, password: string, recaptcha: string } | undefined, setToken: any, setErrorMessagesAux: any, errorMessagesAux: any) => {
    const [errorMessages, setErrorMessages] = [errorMessagesAux, setErrorMessagesAux];
    let success = false;
    if (loginDto === undefined) return false;

    setToken('');

    if (/^[a-zA-Z0-9]+@[a-zA-Z0-9]+\.[A-Za-z]+$/.test(loginDto.login)) {
        const email = loginDto.login;
        const password = loginDto.password;
        const recaptcha = loginDto.recaptcha;

        await axios.post(url + "login", { email, password, recaptcha })
            .then(res => {
                if (res.status === 200) {
                    setToken(res.data);
                    success = true;
                }
            })
            .catch(res => {
                if (res.response.status === 400) {
                    let errors: {
                        type: string,
                        message: string
                    }[] = [];
                    res.response.data.forEach((e: string) => {
                        console.log(e);
                        errors.push({type: "Error", message: e});
                    });
                    setErrorMessages(errors);
                }
                if (res.response.status === 401) {
                    let errors: {
                        type: string,
                        message: string
                    }[] = [];
                    errors.push({ type: 'Authentication', message: 'Email and password do not match' });
                    setErrorMessages(errors);
                }
            });
    }
    else {
        const username = loginDto.login;
        const password = loginDto.password;
        const recaptcha = loginDto.recaptcha;

        await axios.post(url + "login", { username, password, recaptcha })
            .then(res => {
                if (res.status === 200) {
                    setToken(res.data);
                    success = true;
                }
            })
            .catch(res => {
                if (res.response.status === 400) {
                    let errors: {
                        type: string,
                        message: string
                    }[] = [];
                    res.response.data.forEach((e: string) => {
                        console.log(e);
                        errors.push({type: "Error", message: e});
                    });
                    setErrorMessages(errors);
                }
                if (res.response.status === 401) {
                    let errors: {
                        type: string,
                        message: string
                    }[] = [];
                    errors.push({ type: 'Authentication', message: 'Username and password do not match' });
                    setErrorMessages(errors);
                }
            });
    }
    return success;
}

export const register = async (registerDto: { username: string, password: string, email: string, birthDate: Date, recaptcha: string}, setErrorMessagesAux: any, errorMessagesAux: any) => {
    const [errorMessages, setErrorMessages] = [errorMessagesAux, setErrorMessagesAux];
    const username = registerDto.username;
    const email = registerDto.email;
    const password = registerDto.password;
    let bd = registerDto.birthDate;
    let birthDate = Moment(bd).format('YYYY-MM-DD');

    const recaptcha = registerDto.recaptcha;
    let success = false;
    await axios.post(url + 'register', { username, email, password, birthDate, recaptcha })
        .then((res) => {
            success = true;
        })
        .catch(res => {
            if (res.response.status === 201) {
                success = true;
                console.log(success);
            }
            if (res.response.status === 400) {
                let errors: {
                    type: string,
                    message: string
                }[] = [];
                res.response.data.forEach((e: { type: string, message: string }) => {
                    errors.push(e);
                });
                setErrorMessages(errors);
            }
        });
    return success;
}

export const verifyAccount = async (token: any) => {
    let success = false;
    await axios.post(url + "verify_account", {token})
        .then(res => {
            success = true;
        })
        .catch((res) => {
            console.log(res.response.data);
        })
    return success;
}

File: api/src/Domain/Services/Authentication/AuthService.cs
using Microsoft.IdentityModel.Tokens;
using System.IdentityModel.Tokens.Jwt;
using System.Security.Claims;
using System.Text;
using JwtRegisteredClaimNames = Microsoft.IdentityModel.JsonWebTokens.JwtRegisteredClaimNames;
using Mailosaur.Models;
using Mailosaur;
using Microsoft.Extensions.Configuration;
using Core.Entities;
using Core.Abstractions.Services.Auth;
using Core.Abstractions.Context;
using Core.Abstractions.Models;
using System.Security.Authentication;
using FluentValidation;
using Core.Application.Validators;

namespace Core.Services.Authentication
{
    internal class AuthService : IAuthService
    {
        private readonly IUserRepository _userRepository;
        private readonly IUnitOfWork _unitOfWork;
        private readonly RegisterValidator _registerValidator;
        private readonly LoginValidator _loginValidator;
        private readonly IConfiguration _configuration;
        public AuthService(IUserRepository userRepository, IUnitOfWork unitOfWork, RegisterValidator registerValidator, LoginValidator loginValidator, IConfiguration configuration)
        {
            _userRepository = userRepository;
            _unitOfWork = unitOfWork;
            _registerValidator = registerValidator;
            _loginValidator = loginValidator;
            _configuration = configuration;
        }

        public async Task<Guid> LoginUserAsync(IUser loginDto)
        {
            var results = _loginValidator.Validate(loginDto);
            if (results.IsValid == false)
            {
                throw new ValidationException(results.Errors.First().ErrorMessage);
            }
            var user = (User?)_userRepository.GetUser(x => x.Username == loginDto.Username || x.Email == loginDto.Email);
            if (user == null)
            {
                throw new AuthenticationException();
            }
            if (BCrypt.Net.BCrypt.Verify(loginDto.Password, user.Password) == false)
            {
                throw new AuthenticationException();
            }
            if (user.isVerified == false)
            {
                // return "ERROR: Email is not verified.";
            }
            user.LastAuthenticationDate = DateTime.UtcNow;
            await _unitOfWork.SaveAsync();

            var jwtToken = CreateJwtToken(user.ID);
            return user.ID;
        }

        public string CreateJwtToken(Guid id)
        {
            var user = _userRepository.GetUserById(id);
            if(user == null)
            {
                throw new ArgumentNullException(nameof(user));
            }

            var issuer = _configuration["JWT:Issuer"];
            var audience = _configuration["JWT:Audience"];
            var key = Encoding.ASCII.GetBytes(_configuration["JWT:Key"]!);

            var tokenDescriptor = new SecurityTokenDescriptor
            {
                Subject = new ClaimsIdentity(new[]
                {
                    new Claim("id", user.ID.ToString()),
                    new Claim(JwtRegisteredClaimNames.Name, user.Username),
                    new Claim(JwtRegisteredClaimNames.Email, user.Email),
                    new Claim(JwtRegisteredClaimNames.Jti, Guid.NewGuid().ToString())
                }),
                Issuer = issuer,
                Audience = audience,
                SigningCredentials = new SigningCredentials
                (new SymmetricSecurityKey(key), SecurityAlgorithms.HmacSha256Signature),
                Expires = DateTime.UtcNow.AddDays(1)
            };

            var tokenHandler = new JwtSecurityTokenHandler();
            var token = tokenHandler.CreateJwtSecurityToken(tokenDescriptor);
            var jwtToken = tokenHandler.WriteToken(token);

            return jwtToken;
        }

        public async Task<List<IValidationFailureResponse>?> RegisterUserAsync(IUser registerDto)
        {
            //var results = await _registerValidator.ValidateAsync(registerDto);
            //if (results.IsValid == false)
            //{
            //    var errors = new List<ValidationFailureResponse>();
            //    foreach (var error in results.Errors)
            //    {
            //        new ValidationFailureResponse { Type = error.PropertyName, Message = error.ErrorMessage })
            //    return errors;
            //}

            //var user = new User
            //{
            //    Username = registerDto.Username,
            //    Password = BCrypt.Net.BCrypt.HashPassword(registerDto.Password),
            //    Email = registerDto.Email,
            //    BirthDate = registerDto.BirthDate,
            //    AccountCreationDate = DateOnly.FromDateTime(DateTime.UtcNow),
            //    LastAuthenticationDate = null,
            //    isVerified = false
            //};

            //_dataContext.Users.Add(user);
            //await _dataContext.SaveChangesAsync();

            //SendValidationEmail(user);

            //return new Result<bool, List<IValidationFailureResponse>>.Success(true);
            return null;
        }

        public void SendValidationEmail(Guid id)
        {
            var user = _userRepository.GetUserById(id);
            if(user is null)
            {
                throw new ArgumentException();
            }

            var jwtToken = GenerateVerificationJwt(user.ID);

            var clientUrl = _configuration["Client:Url"];
            var webAdress = clientUrl + "/verify_account?token=" + jwtToken;

            var issuer = _configuration["Email:Address"];
            var issuerPassword = _configuration["Email:Password"];

            var apiKey = _configuration["Mailosaur:Api:Key"];
            var serverId = _configuration["Mailosaur:Id"];

            var mailosaur = new MailosaurClient(apiKey);

            mailosaur.Messages.Create(serverId, new MessageCreateOptions()
            {
                To = user.Email,
                Subject = "Email verification - FitnessCat",
                Send = true,
                Html = @"<h1>FitnessCat: New Account Verification</h1>" +
                @"<a href=" + webAdress + ">Please click here to verify your new FitnessCat account</a>"
            });
            //Debug.WriteLine(webAdress);
        }

        public string GenerateVerificationJwt(Guid id)
        {
            var user = _userRepository.GetUserById(id);
            if(user is null)
            {
                throw new ArgumentException();
            }

            var issuer = _configuration["JWT:Issuer"];
            var audience = _configuration["JWT:Audience"];
            var key = Encoding.ASCII.GetBytes(_configuration["JWT:Key"]!);

            var tokenDescriptor = new SecurityTokenDescriptor
            {
                Subject = new ClaimsIdentity(new[]
                {
                    new Claim("Email", user.Email),
                    new Claim(JwtRegisteredClaimNames.Jti, Guid.NewGuid().ToString())
                }),
                Issuer = issuer,
                Audience = audience,
                SigningCredentials = new SigningCredentials
                (new SymmetricSecurityKey(key), SecurityAlgorithms.HmacSha256Signature),
                Expires = DateTime.UtcNow.AddDays(2)
            };

            var tokenHandler = new JwtSecurityTokenHandler();
            var token = tokenHandler.CreateJwtSecurityToken(tokenDescriptor);
            var jwtToken = tokenHandler.WriteToken(token);

            return jwtToken;
        }

        public async Task<Guid> VerifyAccount(IToken token)
        {
            var key = Encoding.ASCII.GetBytes(_configuration["JWT:Key"]!);
            var issuer = _configuration["JWT:Issuer"];
            var audience = _configuration["JWT:Audience"];

            var handler = new JwtSecurityTokenHandler();
            var validations = new TokenValidationParameters
            {
                ValidIssuer = issuer,
                ValidAudience = audience,
                ValidateIssuerSigningKey = true,
                ValidateIssuer = true,
                ValidateAudience = true,
                IssuerSigningKey = new SymmetricSecurityKey(key)
            };
            var claims = handler.ValidateToken(token.Token, validations, out var tokenSecure);
            var claim = claims.Claims.FirstOrDefault(o => o.Type == JwtRegisteredClaimNames.Email);
            if (claim == null)
            {
                throw new InvalidOperationException();
            }

            var email = claim.Value;
            var user = _userRepository.GetUser(user => user.Email == email);
            if (user == null)
            {
                throw new UnauthorizedAccessException();
            }
            if (user.isVerified == false)
            {
                //
            }

            user.isVerified = true;
            await _unitOfWork.SaveAsync();

            return user.ID;
        }

    }
}
